lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -2584,37 +2586,104 @@ lldb::addr_t SwiftLanguageRuntime::GetAsyncContext(RegisterContext *regctx) {
+/// of the async register. This is a simplified version of the methods in
+/// RegisterContextUnwind, since plumbing access to those here would be
+/// challenging.
+std::optional<addr_t>
+ReadAsyncRegFromUnwind(SymbolContext &sc, Process &process, Address pc,
+                       Address func_start_addr, RegisterContext &regctx,
+                       AsyncUnwindRegisterNumbers regnums) {
+  FuncUnwindersSP unwinders =
+      pc.GetModule()->GetUnwindTable().GetFuncUnwindersContainingAddress(pc,
+                                                                         sc);
+  if (!unwinders)
+    return {};
 
-  const uint8_t *expr = triple == llvm::Triple::x86_64
-                            ? g_cfa_dwarf_expression_x86_64
-                            : g_cfa_dwarf_expression_arm64;
-  auto size = triple == llvm::Triple::x86_64
-                  ? sizeof(g_cfa_dwarf_expression_x86_64)
-                  : sizeof(g_cfa_dwarf_expression_arm64);
-  if (with_double_deref)
-    return llvm::ArrayRef<uint8_t>(expr, size);
-  return llvm::ArrayRef<uint8_t>(expr, size - 1);
+  Target &target = process.GetTarget();
+  UnwindPlanSP unwind_plan =
+      unwinders->GetUnwindPlanAtNonCallSite(target, regctx.GetThread());
+  if (!unwind_plan)
+    return {};
+
+  UnwindPlan::RowSP row = unwind_plan->GetRowForFunctionOffset(
+      pc.GetFileAddress() - func_start_addr.GetFileAddress());
+  UnwindPlan::Row::AbstractRegisterLocation regloc;
+
+  // If the plan doesn't have information about the async register, we can use
+  // its current value, as this is a callee saved register.
+  if (!row->GetRegisterInfo(regnums.async_ctx_regnum, regloc))
+    return ReadRegisterAsAddress(regctx, regnums.async_ctx_regnum);
+
+  // Handle the few abstract locations we are likely to encounter.
+  using RestoreType = UnwindPlan::Row::AbstractRegisterLocation::RestoreType;
+  RestoreType loctype = regloc.GetLocationType();
+  switch (loctype) {
+  case RestoreType::same:
+  case RestoreType::inOtherRegister: {
+    unsigned regnum = loctype == RestoreType::same ? regnums.async_ctx_regnum
+                                                   : regloc.GetRegisterNumber();
+    return ReadRegisterAsAddress(regctx, regnum);
+  }
+  case RestoreType::atCFAPlusOffset: {
+    std::optional<addr_t> cfa_addr =
+        GetCFAAddr(process, regctx, row->GetCFAValue());
+    return ReadPtrFromAddr(process, cfa_addr, regloc.GetOffset());
+  }
+  case RestoreType::isCFAPlusOffset: {
+    std::optional<addr_t> cfa_addr =
+        GetCFAAddr(process, regctx, row->GetCFAValue());
+    if (!cfa_addr)
+      return {};
+    return *cfa_addr + regloc.GetOffset();
+  }
+  case RestoreType::isConstant:
+    return regloc.GetConstant();
+  case RestoreType::unspecified:
+  case RestoreType::undefined:
+  case RestoreType::atAFAPlusOffset:
+  case RestoreType::isAFAPlusOffset:
+  case RestoreType::isDWARFExpression:
+  case RestoreType::atDWARFExpression:
+    return {};
```

Comment:
What does it mean if we reach this case?

Reply:
This means it's an abstract register location that this method doesn't handle today.  I'd prob have an lldbassert or assert, I forget which is preferred these days.

Reply:
So I don't want to assert here because we shouldn't be crashing if we fail to produce a virtual backtrace, instead we will just produce a normal backtrace. But maybe we can use Adrian's suggestion of either some kind of `Expected` or logging. I'll give this a try now

Reply:
OK sounds good, I think of assert/lldbassert as "fail in CI and on our desktops with self-built lldb's" and what we're talking about is an unwind plan from the assembly scanner that this method isn't prepared to handle -- so it's going to be a clang change years from now in codegen or lldb's instruction scanner that results in us hitting this.